update_copyright.d

```diff
@@ -0,0 +1,358 @@
+                }
+                else if (line.matchAll(otherCopyrightRe))
+                {
+                    stderr.writeln(loc, ": unrecognised copyright: ", line.strip);
+                    //errors++;
+                }
+                lines.put(line);
+                lines.put('\n');
+            }
+            catch (Exception)
+            {
+                if (this.verbose)
+                    stderr.writeln(file, ": bad input file");
+                errors++;
+                break;
+            }
+        }
+        h.close();
+
+        // If something changed, write the new file out.
+        if (changed && !errors)
+        {
+            auto tmpfile = file ~ ".tmp";
+            h = File(tmpfile, "w");
+            h.write(lines.data);
+            h.close();
+            rename(tmpfile, file);
+        }
+    }
+
+    String stripContinuation(String)(String line)
+    {
+        line = line.stripLeft();
+        auto match = line.matchAll(commentRe);
+        if (match)
+        {
+            auto captures = match.front;
+            line = captures.post.stripLeft();
+        }
+        return line;
+    }
+
+    bool isComplete(Match)(Match match)
+    {
+        auto captures = match.front;
+        return captures.length >= 5 && captures[4] in this.holders;
+    }
+
+    bool updateCopyright(String, Match)(FileLocation loc, ref String line, Match match, ref int errors)
+    {
+        auto captures = match.front;
+        if (captures.length < 5)
+        {
+            stderr.writeln(loc, ": missing copyright holder");
+            errors++;
+            return false;
+        }
+
+        // See if copyright is associated with package author.
+        // Update the author so as to be consistent everywhere.
+        auto holder = captures[4];
+        if (holder !in this.holders)
+        {
+            stderr.writeln(loc, ": unrecognised copyright holder: ", holder);
+            errors++;
+            return false;
+        }
+        else if (!this.holders[holder])
+            return false;
+
+        // Update the copyright years.
+        auto years = captures[2].strip;
+        if (!this.canonicaliseYears(years))
+        {
+            stderr.writeln(loc, ": unrecognised year string: ", years);
+            errors++;
+            return false;
+        }
+
+        // Make sure (C) is present.
+        auto intro = captures[1];
+        if (intro.endsWith("right"))
+            intro ~= " " ~ this.copyright;
+        else if (intro.endsWith("(c)"))
+            intro = intro[0 .. $ - 3] ~ this.copyright;
+
+        // Construct the copyright line, removing any 'by '.
+        auto newLine = captures.pre ~ intro ~ " " ~ years ~ " by " ~ this.author ~ captures.post;
+        if (line != newLine)
+        {
+            line = newLine;
+            return true;
+        }
+        return false;
+    }
+
+    bool canonicaliseYears(String)(ref String years)
```

Comment:
Isn't it ironic that there are two spelling of this word? (-> https://en.wiktionary.org/wiki/canonicalize)
IIIRC American English is used for D.